### PR TITLE
Jitter: use bignum

### PR DESCRIPTION
**Pull request by serpilliere**
Remove uint128_t use.
Miasm now uses bignums to manipulate numbers bigger than 64 bits.
This allows full support of 128/256.. vectors in jitters !

Note:
Jit on 32 bit machine (as i386) should work again with this patch

### CODE REVIEW

File: miasm2/ir/translators/C.py [hunk 1]
```diff
@@ -194,38 +284,123 @@ def from_ExprOp(self, expr):
...
                               "fprem", "fprem_lsb", "fyl2x", "fpatan"]):
-                return "fpu_%s%d(%s, %s)" % (
-                    expr.op,
-                    expr.size,
-                    self.from_expr(expr.args[0]),
-                    self.from_expr(expr.args[1]),
-                )
+                arg0 = self.from_expr(expr.args[0])
+                arg1 = self.from_expr(expr.args[1])
+                if not expr.args[0].size <= self.NATIVE_INT_MAX_SIZE:
```
Comment (commial) [inline miasm2/ir/translators/C.py]: If the argument is > 64 bit, the operation must fail. If I guess correctly that this check is for case when `NATIVE_INT_MAX_SIZE` is less than 64, an additional assert would avoid a silent fail

File: miasm2/ir/translators/C.py [hunk 2]
```diff
@@ -194,38 +284,123 @@ def from_ExprOp(self, expr):
...
+                    out = "bignum_%s(%s, %s, %d)" % (
+                        expr.op,
+                        self.from_expr(expr.args[0]),
+                        self.from_expr(expr.args[1]),
+                        expr.size
+                    )
+                    out = "bignum_mask(%s, %d)"% (out, expr.size)
+                return out
+
+            elif expr.op == "/":
```
Comment (commial) [inline miasm2/ir/translators/C.py]: I'm not sure "/" and "%" are still supported, because there are ambiguous compared to "udiv/idiv" and "umod/imod"

File: miasm2/jitter/JitCore.h [hunk 3]
```diff
@@ -140,14 +207,29 @@ uint8_t MEM_LOOKUP_08(JitCpu* jitcpu, uint64_t addr);
...
 uint32_t MEM_LOOKUP_32(JitCpu* jitcpu, uint64_t addr);
 uint64_t MEM_LOOKUP_64(JitCpu* jitcpu, uint64_t addr);
-uint128_t MEM_LOOKUP_128(JitCpu* jitcpu, uint64_t addr);
+
+bn_t MEM_LOOKUP_BN_BN(JitCpu* jitcpu, int size, bn_t addr);
+bn_t MEM_LOOKUP_INT_BN(JitCpu* jitcpu, int size, uint64_t addr);
+
+uint64_t MEM_LOOKUP_BN_INT(JitCpu* jitcpu, int size, bn_t addr);
+
+//bn_t MEM_LOOKUP_08(JitCpu* jitcpu, bn_t addr);
```
Comment (commial) [inline miasm2/jitter/JitCore.h]: Probably no more wanted

File: miasm2/jitter/Jitllvm.c [hunk 4]
```diff
@@ -6,6 +6,7 @@
 #include "queue.h"
 #include "vm_mngr.h"
 #include "vm_mngr_py.h"
+#include "bn.h"
```
Comment (commial) [inline miasm2/jitter/Jitllvm.c]: JitCore being already imported, is there a need for `bn.h`?

File: miasm2/jitter/bn.h [hunk 5]
```diff
@@ -0,0 +1,141 @@
+#ifndef __BIGNUM_H__
```
Comment (commial) [inline miasm2/jitter/bn.h]: It would be nice to add the source of the original library

File: miasm2/jitter/codegen.py [hunk 6]
```diff
@@ -229,10 +232,16 @@ def gen_c_assignments(self, assignblk):
         for expr, prefetcher in sorted(prefetchers.iteritems()):
             str_src = self.id_to_c(expr)
             str_dst = self.id_to_c(prefetcher)
-            c_prefetch.append('%s = %s;' % (str_dst, str_src))
+            if expr.size <= self.translator.NATIVE_INT_MAX_SIZE:
```
Comment (commial) [inline miasm2/jitter/codegen.py]: There is no diff between if and else

File: miasm2/jitter/codegen.py [hunk 7]
```diff
@@ -244,24 +253,51 @@ def gen_c_assignments(self, assignblk):
...
                 raise ValueError("Unknown dst")
 
         for dst, new_dst in dst_var.iteritems():
             if dst == self.ir_arch.IRDst:
                 continue
-            c_updt.append('%s = %s;' % (self.id_to_c(dst), self.id_to_c(new_dst)))
-            c_var.append("uint%d_t %s;" % (new_dst.size, new_dst))
+
+            if dst.size <= self.translator.NATIVE_INT_MAX_SIZE:
+                c_updt.append('%s = %s;' % (self.id_to_c(dst), self.id_to_c(new_dst)))
```
Comment (commial) [inline miasm2/jitter/codegen.py]: This line can be extracted from the if/else

File: miasm2/jitter/jitcore.py [hunk 8]
```diff
@@ -195,9 +195,10 @@ def run_at(self, cpu, offset, stop_offsets):
                 return offset
 
         # Run the block and update cpu/vmmngr state
-        return self.exec_wrapper(offset, cpu, self.offset_to_jitted_func.data,
-                                 stop_offsets,
-                                 self.options["max_exec_per_call"])
+        ret = self.exec_wrapper(offset, cpu, self.offset_to_jitted_func.data,
```
Comment (commial) [inline miasm2/jitter/jitcore.py]: Unnecessary diff

File: miasm2/jitter/llvmconvert.py [hunk 9]
```diff
@@ -307,25 +323,76 @@ def set_vmcpu(self, lookup_table):
...
+            fc_name = "MEM_LOOKUP_%02d" % size
+            fc_ptr = self.mod.get_global(fc_name)
+            addr_casted = builder.zext(addr, LLVMType.IntType(64))
+            ret = builder.call(
+                fc_ptr, [func.local_vars["jitcpu"],addr_casted]
+            )
+        else:
+            assert size == 128
+
+            fc_name = "MEM_LOOKUP_INT_BN_TO_PTR"
```
Comment (commial) [inline miasm2/jitter/llvmconvert.py]: As the code is quite complicated here, a comment describing the different cast would be welcomed

File: miasm2/ir/translators/C.py (hunk 1 above)
Comment (serpilliere) [reply miasm2/ir/translators/C.py]: Done

File: miasm2/ir/translators/C.py (hunk 2 above)
Comment (serpilliere) [reply miasm2/ir/translators/C.py]: Done
I completely remove those operators in the translator:
There were a last use of those operators in the x86 `aam`.
I added the UDIV8/UMOD8 in the op_semantic which where missing and needed to update `aam`

File: miasm2/jitter/JitCore.h (hunk 3 above)
Comment (serpilliere) [reply miasm2/jitter/JitCore.h]: Done

File: miasm2/jitter/Jitllvm.c (hunk 4 above)
Comment (serpilliere) [reply miasm2/jitter/Jitllvm.c]: Yes it is:
* `Jitcore.h` is needed by jitllvm (for jitcpu strcuts
* as jitcore.h export bn_t struct is need bn_t defnied BEFORE any inclusion
Conclusion:
bn.h is need in each C file using `jitcore.h`

File: miasm2/jitter/bn.h (hunk 5 above)
Comment (serpilliere) [reply miasm2/jitter/bn.h]: Done

File: miasm2/jitter/codegen.py (hunk 6 above)
Comment (serpilliere) [reply miasm2/jitter/codegen.py]: Done

File: miasm2/jitter/codegen.py (hunk 7 above)
Comment (serpilliere) [reply miasm2/jitter/codegen.py]: Done

File: miasm2/jitter/jitcore.py (hunk 8 above)
Comment (serpilliere) [reply miasm2/jitter/jitcore.py]: Done

File: miasm2/jitter/llvmconvert.py [hunk 10]
```diff
@@ -307,25 +323,76 @@ def set_vmcpu(self, lookup_table):
...
-                                   LLVMType.IntType(64))
+        if size <= 64:
+            fc_name = "MEM_LOOKUP_%02d" % size
+            fc_ptr = self.mod.get_global(fc_name)
+            addr_casted = builder.zext(addr, LLVMType.IntType(64))
+            ret = builder.call(
+                fc_ptr, [func.local_vars["jitcpu"],addr_casted]
+            )
+        else:
+            assert size == 128
```
[the remark this answers is not in the document]
Comment (serpilliere) [reply miasm2/jitter/llvmconvert.py]: Done

File: miasm2/jitter/llvmconvert.py (hunk 9 above)
Comment (serpilliere) [reply miasm2/jitter/llvmconvert.py]: Done
